rust/main/chains/hyperlane-starknet/src/ism/aggregation_ism.rs

```diff
@@ -0,0 +1,138 @@
+use cainome::cairo_serde::U256 as StarknetU256;
+use hyperlane_core::{
+    AggregationIsm, ChainResult, ContractLocator, HyperlaneAbi, HyperlaneChain, HyperlaneContract,
+    HyperlaneDomain, HyperlaneMessage, HyperlaneProvider, H256,
+};
+use starknet::accounts::SingleOwnerAccount;
+use starknet::providers::AnyProvider;
+use starknet::signers::LocalWallet;
+use tracing::instrument;
+
+use crate::contracts::aggregation_ism::{
+    AggregationIsm as StarknetAggregationIsmInternal, Bytes as StarknetBytes,
+    Message as StarknetMessage,
+};
+use crate::error::HyperlaneStarknetError;
+use crate::{build_single_owner_account, ConnectionConf, Signer, StarknetProvider};
+
+impl<A> std::fmt::Display for StarknetAggregationIsmInternal<A>
+where
+    A: starknet::accounts::ConnectedAccount + Sync + std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+/// A reference to a AggregationISM contract on some Starknet chain
+#[derive(Debug)]
+#[allow(unused)]
+pub struct StarknetAggregationIsm {
+    contract: Arc<StarknetAggregationIsmInternal<SingleOwnerAccount<AnyProvider, LocalWallet>>>,
+    provider: StarknetProvider,
+    conn: ConnectionConf,
+}
+
+impl StarknetAggregationIsm {
+    /// Create a reference to a AggregationISM at a specific Starknet address on some
+    /// chain
+    pub fn new(
+        conn: &ConnectionConf,
+        locator: &ContractLocator,
+        signer: Signer,
+    ) -> ChainResult<Self> {
+        let account = build_single_owner_account(
+            &conn.url,
+            signer.local_wallet(),
+            &signer.address,
+            false,
+            locator.domain.id(),
+        );
+
+        let contract = StarknetAggregationIsmInternal::new(
+            locator
+                .address
+                .try_into()
+                .map_err(HyperlaneStarknetError::BytesConversionError)?,
+            account,
+        );
+
+        Ok(Self {
+            contract: Arc::new(contract),
+            provider: StarknetProvider::new(locator.domain.clone(), conn),
+            conn: conn.clone(),
+        })
+    }
+
+    #[allow(unused)]
+    pub fn contract(
+        &self,
+    ) -> &StarknetAggregationIsmInternal<SingleOwnerAccount<AnyProvider, LocalWallet>> {
+        &self.contract
+    }
+}
+
+impl HyperlaneChain for StarknetAggregationIsm {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.provider.domain()
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.provider.clone())
+    }
+}
+
+impl HyperlaneContract for StarknetAggregationIsm {
+    fn address(&self) -> H256 {
+        self.contract.address.into()
+    }
+}
+
+#[async_trait]
+impl AggregationIsm for StarknetAggregationIsm {
+    #[instrument(err)]
+    async fn modules_and_threshold(
+        &self,
+        message: &HyperlaneMessage,
+    ) -> ChainResult<(Vec<H256>, u8)> {
+        let message = &StarknetMessage {
```

Comment:
done

note that we need an impl by crate since in the back the abi types generation doesnt merge types accross contracts, this is unfortunate but also allows for more control

Reply:
> since in the back the abi types generation doesnt merge types accross contracts

wah that's pretty crazy, fair enough

Reply:
makes sense why you hadn't implemented the `From` in the first place